### ISSUE TITLE
Diffraction module

### ISSUE THREAD

**Issue by bdice** [glotzerlab/freud#245]
Discussing with @mikemhenry, it would be nice to add a module for computing diffraction patterns in `freud`. There are some existing implementations that could be adapted:
- @cmelab [diffract.py](https://bitbucket.org/cmelab/cme_utils/src/e83c85aed54ee1de388e9c23ce496937da5f1214/cme_utils/analyze/diffract.py?at=master&fileviewer=file-view-default)
- platoviz [Diffraction.py](https://github.com/glotzerlab/platoviz/blob/master/platoviz/modules/Diffraction.py)

**Comment by mikemhenry** [comment]
I'm interested in this!

**Comment by bdice** [comment]
@mikemhenry Awesome! I'm happy to help if you have any questions.

The [Development Guide](https://freud.readthedocs.io/en/stable/development.html) might be a helpful resource. One thing to note: it's perfectly fine to write this as a Cython-only module that doesn't have separate C++ sources, if you want. That should be easier to port from the existing Python codebases.

**Comment by bdice** [comment]
@vyasr What do you think of an API like this? @mikemhenry and I discussed on Slack.

```python
df = freud.diffraction.Diffraction(**some_diffraction_params_tbd)
df.compute(box, points, orientation)
plt.imshow(df.diffraction)
```

@mikemhenry is interested in exposing the 2D-projected data to the user, as well.

**Comment by vyasr** [comment]
I think that's a great idea, I'd love to add diffraction to freud. The examples that you posted above should be a decent guide on how to get started.

The one minor detail about the API that you exposed that I might want to at least think a little bit about is the array ordering. We can discuss this in more detail later, but I do find that the PMFT ordering is very nonintuitive because it's designed around `imshow`. We might want to ask a few other PMFT users about that before making a decision about whether or not to emulate that here.

**Comment by bdice** [comment]
I worked with @alacour to prototype a solution to this issue. I have attached a Jupyter notebook that closely matches a reference implementation. This calculation is "quick and dirty" but probably an alright starting point.

The first computation is the full 3D GaussianDensity, projected to 2D and then FFT'd. The second computation is a 2D projection followed by GaussianDensity (much faster).

![image](https://user-images.githubusercontent.com/3943761/65802112-b27de200-e148-11e9-8f62-3fe0a15f9261.png)

[diffraction-test.zip](https://github.com/glotzerlab/freud/files/3664517/diffraction-test.zip)

**Comment by bdice** [comment]
Some other references:

- LAMMPS `xrd` command: https://lammps.sandia.gov/doc/compute_xrd.html
- `pymatgen.analysis.diffraction` ([source](https://github.com/materialsproject/pymatgen/tree/master/pymatgen/analysis/diffraction), [docs](https://pymatgen.org/pymatgen.analysis.diffraction.xrd.html))

**Comment by klywang** [comment]
@alacour and I tuned the GaussianDensity parameters for a decagonal quasicrystal with results that closely match the expected diffraction (calculated with experimental techniques). I included two files which produce diffractions with 10 (`simulated_ref.gsd`) and 20 (`simulated_15.gsd`) expected peaks:
For `simulated_ref.gsd`
GaussianDensity calculation, then 2D projection
![image](https://user-images.githubusercontent.com/47036428/68508435-69877600-023c-11ea-9375-e0131f13427b.png)

2D projection, then GaussianDensity calculation
![image](https://user-images.githubusercontent.com/47036428/68508445-7015ed80-023c-11ea-89c1-b826498796ff.png)

Expected:
![image](https://user-images.githubusercontent.com/47036428/68508884-96885880-023d-11ea-83ee-35bd823bc6bf.png)


For `simulated_15.gsd`:
2D projection, then GaussianDensity calculation
![image](https://user-images.githubusercontent.com/47036428/68508477-80c66380-023c-11ea-9729-af81cf1cc9bb.png)

Expected:
![image](https://user-images.githubusercontent.com/47036428/68508889-9c7e3980-023d-11ea-806c-f247fe056ac4.png)


These seem to show very good agreement with the reference diffractions, calculated using standard image analysis based techniques.

It appears the original sample file was not the best crystal for analysis. The parameters were set so that the size of the pixle was significantly smaller than the number of particles.

A folder with the updated Jupyter notebook and files are included in [diffraction-test.zip](https://github.com/glotzerlab/freud/files/3826166/diffraction-test.zip).

**Comment by alacour** [comment]
Kwang and I attempted to improve the prototype of the code in jupyter notebook attached above. I have attached it in 
[diffraction-test.zip](https://github.com/glotzerlab/freud/files/3881623/diffraction-test.zip)


Some conclusions/notes we arrived at: 
1) Using Freud's gaussian density on the 3D particle positions and then project the 3D density to 2D worked quickly at rcut's below about 5 * sigma.

2) After taking the log of the intensities produced from the fft, a good vmin and vmax for visualising the peaks was the average of the log intensity and the max of the log intensity, respectively.

3) For an FCC lattice structure and HCP structure, we see lines between the point in the pattern. We believe these to be wrong.

4) We need to address noncubic boxes. Specifically, I need to determine how Gaussian density handles them.
